File: java/base/src/main/java/cn/wjybxx/base/io/PoolableObjectHandlers.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.base.io;

import cn.wjybxx.base.pool.ObjectPool;

import javax.annotation.Nullable;
import java.util.Objects;
import java.util.function.Consumer;
import java.util.function.IntFunction;
import java.util.function.Predicate;
import java.util.function.Supplier;

/**
 * @author wjybxx
 * date - 2024/7/19
 */
public class PoolableObjectHandlers {

    public static <T> PoolableObjectHandler<T> of(Supplier<? extends T> factory,
                                                  @Nullable Consumer<? super T> resetHandler) {
        return new PoolableObjectHandler1<>(factory, resetHandler, null);
    }

    public static <T> PoolableObjectHandler<T> of(Supplier<? extends T> factory,
                                                  @Nullable Consumer<? super T> resetHandler,
                                                  @Nullable Predicate<? super T> filter) {
        return new PoolableObjectHandler1<>(factory, resetHandler, filter);
    }

    public static <T> PoolableObjectHandler<T> of(IntFunction<? extends T> factory,
                                                  @Nullable Consumer<? super T> resetHandler) {
        return new PoolableObjectHandler2<>(factory, resetHandler, null);
    }


    public static <T> PoolableObjectHandler<T> of(IntFunction<? extends T> factory,
                                                  @Nullable Consumer<? super T> resetHandler,
                                                  @Nullable Predicate<? super T> filter) {
        return new PoolableObjectHandler2<>(factory, resetHandler, filter);
    }

    public static PoolableObjectHandler<StringBuilder> newStringBuilderHandler(int minCapacity, int maxCapacity) {
        return new StringBuilderHandler(minCapacity, maxCapacity);
    }

    // region impl

    private static class PoolableObjectHandler1<T> implements PoolableObjectHandler<T> {

        private final Supplier<? extends T> factory;
        private final Consumer<? super T> resetHandler;
        private final Predicate<? super T> filter;

        public PoolableObjectHandler1(Supplier<? extends T> factory,
                                      Consumer<? super T> resetHandler,
                                      Predicate<? super T> filter) {
            this.factory = Objects.requireNonNull(factory, "factory");
            this.resetHandler = resetHandler;
            this.filter = filter;
        }

        @Override
        public T create(ObjectPool<? super T> pool, int capacity) {
            return factory.get();
        }

        @Override
        public boolean test(T obj) {
            return filter == null || filter.test(obj);
        }

        @Override
        public void reset(T obj) {
            if (resetHandler != null) {
                resetHandler.accept(obj);
            }
        }

        @Override
        public void destroy(T obj) {

        }
    }

    private static class PoolableObjectHandler2<T> implements PoolableObjectHandler<T> {

        private final IntFunction<? extends T> factory;
        private final Consumer<? super T> resetHandler;
        private final Predicate<? super T> filter;

        public PoolableObjectHandler2(IntFunction<? extends T> factory,
                                      Consumer<? super T> resetHandler,
                                      Predicate<? super T> filter) {
            this.factory = Objects.requireNonNull(factory, "factory");
            this.resetHandler = resetHandler;
            this.filter = filter;
        }

        @Override
        public T create(ObjectPool<? super T> pool, int capacity) {
            return factory.apply(capacity);
        }

        @Override
        public boolean test(T obj) {
            return filter == null || filter.test(obj);
        }

        @Override
        public void reset(T obj) {
            if (resetHandler != null) {
                resetHandler.accept(obj);
            }
        }

        @Override
        public void destroy(T obj) {

        }
    }

    private static class StringBuilderHandler implements PoolableObjectHandler<StringBuilder> {

        final int minCapacity;
        final int maxCapacity;

        StringBuilderHandler(int minCapacity, int maxCapacity) {
            if (minCapacity < 0 || maxCapacity < minCapacity) {
                throw new IllegalArgumentException();
            }
            this.minCapacity = minCapacity;
            this.maxCapacity = maxCapacity;
        }

        @Override
        public StringBuilder create(ObjectPool<? super StringBuilder> pool, int capacity) {
            return new StringBuilder(capacity > 0 ? capacity : minCapacity);
        }

        @Override
        public boolean test(StringBuilder obj) {
            return obj.capacity() >= minCapacity && obj.capacity() <= maxCapacity;
        }

        @Override
        public void reset(StringBuilder obj) {
            obj.setLength(0);
        }

        @Override
        public void destroy(StringBuilder obj) {

        }
    }
    // endregion
}

File: csharp/Wjybxx.Commons.Core/src/IO/PoolableObjectHandlers.cs
#region LICENSE

// Copyright 2024 wjybxx(845740757@qq.com)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System;
using System.Text;
using Wjybxx.Commons.Pool;

namespace Wjybxx.Commons.IO;

/// <summary>
/// 提供基础的handler实现
/// </summary>
public static class PoolableObjectHandlers
{
    /// <summary>
    /// 
    /// </summary>
    /// <param name="factory">工厂方法</param>
    /// <param name="resetHandler">重置方法</param>
    /// <param name="filter">过滤器</param>
    /// <returns></returns>
    public static IPoolableObjectHandler<T> Of<T>(Func<T> factory,
                                                  Action<T>? resetHandler = null,
                                                  Func<T, bool>? filter = null) {
        return new PoolableObjectHandler1<T>(factory, resetHandler, filter);
    }

    /// <summary>
    /// 
    /// </summary>
    /// <param name="factory">工厂方法</param>
    /// <param name="resetHandler">重置方法</param>
    /// <param name="filter">过滤器</param>
    /// <returns></returns>
    public static IPoolableObjectHandler<T> Of<T>(Func<int, T> factory,
                                                  Action<T>? resetHandler = null,
                                                  Func<T, bool>? filter = null) {
        return new PoolableObjectHandler2<T>(factory, resetHandler, filter);
    }

    /// <summary>
    /// 
    /// </summary>
    /// <param name="minCapacity">初始空间</param>
    /// <param name="maxCapacity">最大空间</param>
    /// <returns></returns>
    public static IPoolableObjectHandler<StringBuilder> NewStringBuilderHandler(int minCapacity, int maxCapacity) {
        return new StringBuilderHandler(minCapacity, maxCapacity);
    }

    private class PoolableObjectHandler1<T> : IPoolableObjectHandler<T>
    {
        private readonly Func<T> _factory;
        private readonly Action<T>? _resetHandler;
        private readonly Func<T, bool>? _filter;

        internal PoolableObjectHandler1(Func<T> factory, Action<T>? resetHandler = null, Func<T, bool>? filter = null) {
            _factory = factory ?? throw new ArgumentNullException(nameof(factory));
            _resetHandler = resetHandler;
            _filter = filter;
        }

        public T Create(IObjectPool<T> pool, int capacity) {
            return _factory.Invoke();
        }

        public bool Test(T obj) {
            return _filter == null || _filter.Invoke(obj);
        }

        public void Reset(T obj) {
            if (_resetHandler != null) {
                _resetHandler.Invoke(obj);
            }
        }

        public void Destroy(T obj) {
        }
    }

    private class PoolableObjectHandler2<T> : IPoolableObjectHandler<T>
    {
        private readonly Func<int, T> _factory;
        private readonly Action<T>? _resetHandler;
        private readonly Func<T, bool>? _filter;

        internal PoolableObjectHandler2(Func<int, T> factory, Action<T>? resetHandler = null, Func<T, bool>? filter = null) {
            _factory = factory ?? throw new ArgumentNullException(nameof(factory));
            _resetHandler = resetHandler;
            _filter = filter;
        }

        public T Create(IObjectPool<T> pool, int capacity) {
            return _factory.Invoke(capacity);
        }

        public bool Test(T obj) {
            return _filter == null || _filter.Invoke(obj);
        }

        public void Reset(T obj) {
            if (_resetHandler != null) {
                _resetHandler.Invoke(obj);
            }
        }

        public void Destroy(T obj) {
        }
    }

    private class StringBuilderHandler : IPoolableObjectHandler<StringBuilder>
    {
        private readonly int minCapacity;
        private readonly int maxCapacity;

        internal StringBuilderHandler(int minCapacity, int maxCapacity) {
            if (minCapacity < 0 || maxCapacity < minCapacity) {
                throw new ArgumentException();
            }
            this.minCapacity = minCapacity;
            this.maxCapacity = maxCapacity;
        }

        public StringBuilder Create(IObjectPool<StringBuilder> pool, int capacity) {
            return new StringBuilder(capacity > 0 ? capacity : minCapacity);
        }

        public bool Test(StringBuilder obj) {
            return obj.Capacity >= minCapacity && obj.Capacity <= maxCapacity;
        }

        public void Reset(StringBuilder obj) {
            obj.Clear();
        }

        public void Destroy(StringBuilder obj) {
        }
    }
}